topic-operator/src/main/java/io/strimzi/operator/topic/KafkaStreamsConfiguration.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.topic;
+
+import io.apicurio.registry.streams.diservice.AsyncBiFunctionService;
+import io.apicurio.registry.streams.diservice.AsyncBiFunctionServiceGrpcLocalDispatcher;
+import io.apicurio.registry.streams.diservice.DefaultGrpcChannelProvider;
+import io.apicurio.registry.streams.diservice.DistributedAsyncBiFunctionService;
+import io.apicurio.registry.streams.diservice.LocalService;
+import io.apicurio.registry.streams.diservice.proto.AsyncBiFunctionServiceGrpc;
+import io.apicurio.registry.streams.distore.DistributedReadOnlyKeyValueStore;
+import io.apicurio.registry.streams.distore.FilterPredicate;
+import io.apicurio.registry.streams.distore.KeyValueSerde;
+import io.apicurio.registry.streams.distore.KeyValueStoreGrpcImplLocalDispatcher;
+import io.apicurio.registry.streams.distore.UnknownStatusDescriptionInterceptor;
+import io.apicurio.registry.streams.distore.proto.KeyValueStoreGrpc;
+import io.apicurio.registry.streams.utils.ForeachActionDispatcher;
+import io.apicurio.registry.streams.utils.Lifecycle;
+import io.apicurio.registry.streams.utils.LoggingStateRestoreListener;
+import io.apicurio.registry.utils.ConcurrentUtil;
+import io.apicurio.registry.utils.kafka.AsyncProducer;
+import io.apicurio.registry.utils.kafka.ProducerActions;
+import io.grpc.Server;
+import io.grpc.ServerBuilder;
+import io.grpc.ServerInterceptors;
+import io.grpc.Status;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.errors.InvalidStateStoreException;
+import org.apache.kafka.streams.state.HostInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Configuration required for KafkaStreamsTopicStore
+ */
+public class KafkaStreamsConfiguration {
+    private static final Logger log = LoggerFactory.getLogger(KafkaStreamsConfiguration.class);
+
+    private final List<AutoCloseable> closeables = new ArrayList<>();
+
+    /* test */ KafkaStreams streams;
+    private TopicStore topicStore;
+
+    public void start(Config config, Properties kafkaProperties) {
+        try {
+            String storeTopic = config.get(Config.STORE_TOPIC);
+            String storeName = config.get(Config.STORE_NAME);
+
```

Comment:
We should probably validate the topic at this point. We don't really want to rely on autocreation because :

* That will eventually be removed.
* It's quite easy to end up with a misconfigured topic that way (e.g. if default RF=1).

It's a little tricky because the TO needs to work in clusters with only 1 broker, but in clusters with 3 or more we'd want `RF=3` and `min.insync.replicas=2`. You should be able to use the Kafka `Admin` client something like this:

0. Describe the cluster.
1. Check whether the topic exists:
2. If the topic does not exist try to create it with `RF=min(3, clusterSize)` and `minISR=RF-1`. If created OK we're good. If not (e.g. authorization) then log an error telling the user to create it and exit.
3. If the topic already exists check whether `RF=min(3, clusterSize) && minISR=RF-1`. If it does we're good. If not log a warning that the durability of the topic is not sufficient for production use and proceed.

Reply:
What do you mean under "describe the cluster"?
I guess AdminClient has some API for this?

Reply:
Yes, `describeCluster()`, `describeTopics()` and `describeConfigs()` are all probably relevant to doing this.

Reply:
> If not (e.g. authorization) then log an error telling the user to create it and exit.

@tombentley out of curiosity, doesn't the TO uses a super-user for doing that anyway?

Reply:
@tombentley how many partitions?
And how do I set minISR when creating a new topic?

Reply:
@ppatierno is does when it's deployed by the CO, but there are users who run it stand alone, with Kafka clusters not managed by the CO, so we can't assume we have authz.

Reply:
@alesj [min.insync.replicas](https://kafka.apache.org/documentation.html#min.insync.replicas) is a topic config, so `NewTopic(...).configs(topicConfigs)` will do it (yeah, the fact there there's not a constructor taking the configs makes it a little less discoverable).

Reply:
tbh I am not sure how many people are using TO in this way but I have no clue otherwise I would remove the support for it.

Reply:
BTH @ppatierno supporting that use case adds very little additional complexity. Removing it would only serve to have fewer users, not simplify anything significantly.

Reply:
I know, it's just about a warning :-)